### PR TITLE
Properly close notification substreams

### LINKED ISSUE #8531: Implement proper notification substreams shutdown
In `sc_network/src/protocol/notifications/handler.rs`, we currently don't properly close substreams.
Any close request instantly drops the substream, without calling `poll_close` first.

This is an artefact from when the legacy substream still existed. It was too difficult to actually properly close substreams.

### PR DESCRIPTION
**Pull request by tomaka**
Fix #8531 

This PR properly implements closing notification substreams. This wasn't the case before due to the difficulty in doing so when we still had the legacy substream.

Will start a burn-in.

### CODE REVIEW

**Comment by tomaka** [comment]
Burn-in shows that things are working. However I want to let it run longer, in order to detect potential debug assertion failures.

Since I'm the only networking team members left, I also want to review this code tomorrow or later this week (when I will have forgotten about it).

**Comment by tomaka** [comment]
> assertion failed: in_substream.is_some() || out_substream.is_some()

Line 704

**Comment by tomaka** [comment]
Re-reviewed, and looks good to me.

**Comment by tomaka** [comment]
@ordian @eskimor 
I don't really know who to ask for a review :grimacing: 
As mentioned, I've re-read these changes a couple times (twice in details), and it's gone through a burn-in.
I'm not necessarily asking for a thorough review, but I can't merge this me alone.

File: client/network/src/protocol/notifications/handler.rs [hunk 1]
```diff
@@ -159,6 +159,16 @@ enum State {
 	Closed {
 		/// True if an outgoing substream is still in the process of being opened.
 		pending_opening: bool,
+
+		/// Outbound substream that has been accepted by the remote. Being closed.
+		out_substream_closing: Option<NotificationsOutSubstream<NegotiatedSubstream>>,
```
[the remark this answers is not in the document]
Comment (eskimor) [reply client/network/src/protocol/notifications/handler.rs:164]: Ok, I guess the other questions are rather similar. I think I need some context of this state machine to make sense of the code.

[the remark this answers is not in the document]
Comment (tomaka) [reply client/network/src/protocol/notifications/handler.rs:164]: The enum variant `Closed` represents the state the state machine exposes to the outside (i.e. the behaviour).
When the behaviour asks for closing, we instantaneously respond "ok" and switch to `Closed`. There is "closing" state internally, but we keep the existing substreams in order to close them properly.

There is a delicate situation to handle: the behaviour might want to close, then open, then close, then open, then close, then open, etc. and the handler would need to respond to all these requests one by one.
In order to avoid such a situation, it is illegal for the behaviour to try open when a previous open request hasn't been confirmed yet. However, in order to not make the implementation of the behaviour too complex, it is always legal to close whatever the situation.
This is implemented in the handler by immediately responding to all close requests.

In other words, the API is designed in order to spread the implementation complexity between the behaviour and the handler.

[the remark this answers is not in the document]
Comment (eskimor) [reply client/network/src/protocol/notifications/handler.rs:164]: Thanks Pierre, that helps!

**Review by eskimor — APPROVED** [review]
I guess, it is fine. Is there any way we can make this code simpler at some point?

File: client/network/src/protocol/notifications/handler.rs [hunk 2]
```diff
@@ -509,19 +563,6 @@ impl ProtocolsHandler for NotifsHandler {
 				// to do.
 				return;
 			},
-			State::Opening { ref mut in_substream, .. } |
-			State::Open { ref mut in_substream, .. } => {
-				if in_substream.is_some() {
```
[the remark this answers is not in the document]
Comment (tomaka) [reply client/network/src/protocol/notifications/handler.rs:514]: I moved these two match variants above.

If `in_substream` is `Some`, then we enter the block line 557 where we discard the new substream, because it's illegal to open a second one.

[the remark this answers is not in the document]
Comment (eskimor) [reply client/network/src/protocol/notifications/handler.rs:514]: uh, yeah - I missed that. :-)

**Comment by tomaka** [comment]
> Is there any way we can make this code simpler at some point?

:grimacing: 

The problem is that the logic itself is kind of complicated.

For performance reasons, we have the point that coordinates the whole network (the behaviour) and the point that handles individual connections living in two different tasks and running asynchronously.
I did a lot of brainstorming when designing the way these two components talk to each other, and I don't think it's really possible to make significantly more simple.

**Comment by tomaka** [comment]
Thanks for the review!

**Comment by tomaka** [comment]
bot merge
